### ISSUE TITLE
How to submit labels to AI platform training jobs?

### ISSUE THREAD

**Issue by axeltidemann** [tensorflow/tfx#4385]
**System information**

- Have I specified the code to reproduce the issue (Yes):
- Environment in which the code is executed (e.g., Local(Linux/MacOS/Windows),
  Interactive Notebook, Google Cloud, etc): Google Cloud
- TensorFlow version: 2.5.1
- TFX Version: 1.2.0
- Python version: 3.7.10
- Python dependencies (from `pip freeze` output):

**Describe the current behavior**

I cannot add labels to AI platform training jobs.

**Describe the expected behavior**

According to https://cloud.google.com/ai-platform/training/docs/resource-labels#add-labels, it seems like I should be able to add labels to an AI platform training job like this:

```
tfx.extensions.google_cloud_ai_platform.Trainer(...,
    custom_config = { tfx.extensions.google_cloud_ai_platform.TRAINING_ARGS_KEY: 
       { "scaleTier": ...,
         "labels": {"team": "team-name" }
       }
    }
)
```

**Standalone code to reproduce the issue**

However, the above will fail with the following error message:

```
Invalid JSON payload received. Unknown name "labels" at 'job.training_input': Cannot find field.
```

Which, according to https://cloud.google.com/ai-platform/training/docs/reference/rest/v1/projects.jobs#traininginput also makes sense. I am a bit at a loss, but I am sure it is something obvious.

**Comment by axeltidemann** [comment]
[Here](https://github.com/tensorflow/tfx/blob/421501e3516fbc1b53444e7b790e61c136c90e96/tfx/extensions/google_cloud_ai_platform/runner.py#L153) it actually says: 
```
# TODO(b/168926785): Consider to change executor_class_path to job_labels.
```

**Comment by axeltidemann** [comment]
The above `Labeller` custom component does not work for billing purposes, because `labels` must be set when creating the job on the AI Platform. I did not realise this until now, because there was a mixture of notebooks and jobs with the same labels that masqueraded the issue. @1025KB any thoughts on this?

**Comment by axeltidemann** [comment]
I am trying to switch to `Vertex pipelines` to resolve this issue.

**Comment by axeltidemann** [comment]
Vertex allows the proper use of labels. However, to use existing pipelines, I have made the necessary change (a one-liner) and created a custom TFX image for the AI Platform jobs. This will allow submission of jobs with labels to the AI Platform like so:
```
trainer_args["custom_config"] = { 
  tfx.extensions.google_cloud_ai_platform.TRAINING_ARGS_KEY : {
    "scaleTier": "BASIC_GPU",
    "project": GOOGLE_CLOUD_PROJECT,
    "region": GOOGLE_CLOUD_REGION,
    "masterConfig": {"imageUri": PIPELINE_IMAGE},
    "labels": { "team": "awesome", "project": "super-cool" } 
  }
}
```
However, this technically breaks the `TrainingInput` resource, and the AI Platform will refuse the `labels` key (which I posted [above](https://github.com/tensorflow/tfx/issues/4385#issue-1030087307)). The fix is to remove the `labels` from the `TrainingInput` dictionary and add it to the `job_labels` dictionary in [tfx.extensions.google_cloud_ai_platform.training_clients.py](https://github.com/tensorflow/tfx/blob/421501e3516fbc1b53444e7b790e61c136c90e96/tfx/extensions/google_cloud_ai_platform/training_clients.py#L229):
```
with telemetry_utils.scoped_labels(
    {telemetry_utils.LABEL_TFX_EXECUTOR: executor_class_path}):
  job_labels = telemetry_utils.make_labels_dict()
job_labels.update(training_inputs.pop('labels', {})) # Allows for propagation of labels to AI Platform.
```
The [recommendation on how to contribute to TFX](https://github.com/tensorflow/tfx/blob/beff82fb552ba7f275541f8e2fa9549ad32735a1/CONTRIBUTING.md#merging-external-contributions) says reach out to TFX engineers first to agree on design. What do you think @jiyongjung0 @1025KB, should this be a pull request?

**Comment by 1025KB** [comment]
For Vertex, labels can be [passed](https://github.com/tensorflow/tfx/blob/master/tfx/extensions/google_cloud_ai_platform/training_clients.py#L374) in, for CAIP, it's not [supported](https://github.com/tensorflow/tfx/blob/master/tfx/extensions/google_cloud_ai_platform/training_clients.py#L229) .

When I check the CAIP API, seems label is not mentioned in public doc 
https://cloud.google.com/ai-platform/training/docs/training-jobs

**Comment by axeltidemann** [comment]
If you were to submit a [Job](https://cloud.google.com/ai-platform/training/docs/reference/rest/v1/projects.jobs#Job) instead of TrainingInput, you can provide labels to CAIP. So labels are supported, but just not through the CAIPJobClient.

**Comment by 1025KB** [comment]
I see, seems labels is out of training_input, not at the same level,
doing "job_labels.update(training_inputs.pop('labels', {}))" might cause some confusion
Considering create a top level experimental args (comment it's CAIP only) to pass labels in

**Comment by axeltidemann** [comment]
@1025KB the [VertexJobClient](https://github.com/tensorflow/tfx/blob/421501e3516fbc1b53444e7b790e61c136c90e96/tfx/extensions/google_cloud_ai_platform/training_clients.py#L328) allows for this, it accepts CustomJob (equivalent to Job for the CAIPJobClient) _and_ CustomJobSpec (equivalent to TrainingInput for the CAIPJobClient, however this is deprecated). 

To be clear: you prefer implementing that the CAIPJobClient accepts a Job as well as a TrainingInput resource, like the VertexJobClient does?

**Comment by 1025KB** [comment]
In this [config](https://cloud.google.com/ai-platform/training/docs/reference/rest/v1/projects.jobs#Job) , labels and training input are separate items in config.
{
labels: {}
training_input:{}
}
instead of 
{
training_input: {label:{}}
}

So I think it would be good to keep this [training_input](https://github.com/tensorflow/tfx/blob/94b554b9a36a178276e5839f543824a7f2b2459f/tfx/extensions/google_cloud_ai_platform/training_clients.py#L211) align with cloud's trainining_input.

Consider pop labels [here](https://github.com/tensorflow/tfx/blob/94b554b9a36a178276e5839f543824a7f2b2459f/tfx/extensions/google_cloud_ai_platform/runner.py#L191)  (or even has it as a top level args in trainer's custom_config?)

**Comment by axeltidemann** [comment]
I agree, the proper way would be to change the input argument to the `Trainer` custom config following the config you linked to (and as I referred to above, also known as a Job resource). However, that would break everyone's implementation that use _just_ the TrainingInput. I suspect it is for this reason the VertexJobClient accepts both. 

Would it not be best to support both [Job](https://cloud.google.com/ai-platform/training/docs/reference/rest/v1/projects.jobs#Job) and [TrainingInput](https://cloud.google.com/ai-platform/training/docs/reference/rest/v1/projects.jobs#traininginput) in the CAIPJobClient? Or do you prefer doing it properly, and making it a breaking change in the release of the next TFX version?

**Comment by 1025KB** [comment]
breaking change is not good.
Currently Job config is [created](https://github.com/tensorflow/tfx/blob/master/tfx/extensions/google_cloud_ai_platform/training_clients.py#L235) here.
both job_id and training_input is from an item in Trainer's custom_config, 
so you can add an item for labels in Trainer's custom_config too

**Comment by axeltidemann** [comment]
That would work, it would only require to pop the label [here](https://github.com/tensorflow/tfx/blob/94b554b9a36a178276e5839f543824a7f2b2459f/tfx/extensions/google_cloud_ai_platform/runner.py#L191) (like you suggested), and then update the labels in the `training_job` dictionary [here](https://github.com/tensorflow/tfx/blob/94b554b9a36a178276e5839f543824a7f2b2459f/tfx/extensions/google_cloud_ai_platform/runner.py#L198) (a few lines below), like so:
```
labels = job_args.pop('labels', {}))
....
training_job.update('labels', labels)
```
Then you can call the Trainer like this:
```
tfx.extensions.google_cloud_ai_platform.Trainer(...,
    custom_config = { tfx.extensions.google_cloud_ai_platform.TRAINING_ARGS_KEY: { "scaleTier": ...},
                      "labels": {"team": "awesome" }
    }
)
```
Do you agree @1025KB ? I am happy to create a pull request based on this.

**Comment by 1025KB** [comment]
sounds good to me

**Comment by axeltidemann** [comment]
I have created a pull request, see link above.

**Comment by pindinagesh** [comment]
Hi @axeltidemann 

Could you please move this issue to closed status, since PR is merged. Thank you!

**Comment by axeltidemann** [comment]
@pindinagesh with pleasure.
